### PR TITLE
Store Bulk operations in Redis storage - draft

### PR DESCRIPTION
**Pull request by pmarjan**
##Usage

To use please add to your env.php the following configuration
`'queue_operations' => [
        'save' => 'redis',
        'redis' => [
            'host' => '127.0.0.1',
            'port' => '6379',
            'user' => 'our-redis-user',
            'password' => 'your-auth-pass',
            'persistent_identifier' => '',
            'database' => 'required-database',
            'timeout' => '2.5'
        ]
    ]`

Not present 'queue_operations' or 'save' => 'db', defaults to MySQL usage

##Key details

Any model/repository classes that utilizes EntityManager (in order to call its load() and save() for example), will need to inject `Magento\AsynchronousOperationsRedis\EntityManager\EntityManagerFactory` and call `$entityManagerFactory->create()`
This factory will set `$this->entityManager` to either `Magento\AsynchronousOperationsRedis\EntityManager\EntityManager` which will operate against redis, or to `Magento\Framework\EntityManager\EntityManager` which is the default implementation that operates against mysql.

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/async-import?pullRequest=104) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/async-import?pullRequest=104) before we can accept your contribution.<br/>**4** out of **5** committers have signed the CLA.<br/><br/>:white_check_mark: nuzil<br/>:white_check_mark: pmarjan<br/>:white_check_mark: vkublytskyi<br/>:white_check_mark: Stepa4man<br/>:x: Vasiliev.A<hr/>**Vasiliev.A** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

**Comment by nuzil** [comment]
I made a performance test, and one single request we can see about 5% of performance improvements. 

And if we are talking about 100 items in a bulk, 21% of performance improvements.

Screen Shot 2019-07-26 at 21 40 51
Screen Shot 2019-07-26 at 21 40 44

File: app/code/Magento/AsynchronousOperations/Api/Data/BulkSummaryInterface.php [hunk 1]
```diff
@@ -11,10 +11,12 @@
  * @api
  * @since 100.2.0
  */
-interface BulkSummaryInterface extends \Magento\Framework\Bulk\BulkSummaryInterface
+interface BulkSummaryInterface extends \Magento\Framework\Bulk\BulkSummaryInterface, \Magento\AsynchronousOperationsRedis\Api\RedisIdentityInterface
 {
     const USER_TYPE = 'user_type';
 
+    const REDIS_PREFIX = 'asynchronous.operations.bulk.summary';
```
Comment (nuzil) [inline app/code/Magento/AsynchronousOperations/Api/Data/BulkSummaryInterface.php]: Please move Redis constants to AsynchronousOperationRedis module

File: app/code/Magento/AsynchronousOperations/Api/Data/OperationInterface.php [hunk 2]
```diff
@@ -11,8 +11,10 @@
  * @api
  * @since 100.2.0
  */
-interface OperationInterface extends \Magento\Framework\Bulk\OperationInterface
+interface OperationInterface extends \Magento\Framework\Bulk\OperationInterface, \Magento\AsynchronousOperationsRedis\Api\RedisIdentityInterface
 {
+    const REDIS_PREFIX = 'asynchronous.operations.operation';
```
Comment (nuzil) [inline app/code/Magento/AsynchronousOperations/Api/Data/OperationInterface.php]: Please move Redis constants to AsynchronousOperationRedis module

File: app/code/Magento/AsynchronousOperations/Api/Data/BulkSummaryInterface.php [hunk 3]
```diff
@@ -11,10 +11,12 @@
  * @api
  * @since 100.2.0
  */
-interface BulkSummaryInterface extends \Magento\Framework\Bulk\BulkSummaryInterface
+interface BulkSummaryInterface extends \Magento\Framework\Bulk\BulkSummaryInterface, \Magento\AsynchronousOperationsRedis\Api\RedisIdentityInterface
```
Comment (nuzil) [inline app/code/Magento/AsynchronousOperations/Api/Data/BulkSummaryInterface.php]: Ideally will be build a AsynchronousOperationsRedis module so, that no other modules has a dependency on it, so we can easily remove module and system will proceed to work without dropping an errors.
Lets try to remove dependency and go more to Service Isolation approach

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/async-import?pullRequest=104) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/async-import?pullRequest=104) before we can accept your contribution.<br/>**3** out of **5** committers have signed the CLA.<br/><br/>:white_check_mark: nuzil<br/>:white_check_mark: pmarjan<br/>:white_check_mark: Stepa4man<br/>:x: Vasiliev.A<br/>:x: Volodymyr Kublytskyi<hr/>**Vasiliev.A, Volodymyr Kublytskyi** seem not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).

File: app/code/Magento/AsynchronousOperationsRedis/KeyManager/Key/AbstractKey.php [hunk 4]
```diff
@@ -18,25 +19,46 @@ abstract class AbstractKey implements RedisKeyInterface
...
-        return $entity->getRedisKey();
+        /** @var string $identifier */
+        $identifier = '';
+
+        if ($entityConfig['identifier'] == 'uuid') {
+            $identifier = $entity->getData('uuid') ? $entity->getData('uuid') : $entity->getData('bulk_uuid');
+        }
+
+        if ($entityConfig['identifier'] == 'serialized_data') {
+            $identifier = $this->encryptor->hash($entity->getData('serialized_data'), Encryptor::HASH_VERSION_SHA256);
```
Comment (pmarjan) [inline app/code/Magento/AsynchronousOperationsRedis/KeyManager/Key/AbstractKey.php]: Alex, with this line the key identifier for the Operation model is made based on its serialized data. Please, perhaps suggest correction here?

**Comment by pmarjan** [comment]
The performance differences when focusing directly on the operation that each entity manager does - the one that works with mysql and the one that works with redis.

Measured is the execution of a line like
`$entityManager->save($entity);`

The time in the charts is the total time for a batch of 100 products. Ten separate measurements are made.
The `save` and `delete` are separated, the time for `load` and `update` are summed.

delete-comparison-batch-size-100
load-and-update-comparison-batch-size-100
save-comparison-batch-size-100

**Comment by naydav** [comment]
@nuzil
Please move this PR to https://github.com/magento/magento2/pulls if it's still actual
Also, need to move the corresponding ticket to https://github.com/magento/community-features for a feature request
or in https://github.com/magento/magento2/issues for a bug
Thanks

**Pull request closed without merging**